File: leetcode/python/124_binary_tree_maximum_path_sum.py
class TreeNode:

    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    def maxPathSum(self, root: TreeNode) -> int:
        self.res = float('-inf')

        def get_val(node):
            if not node:
                return 0
            left = max(0, get_val(node.left))
            right = max(0, get_val(node.right))
            self.res = max(self.res, node.val + left + right)
            return node.val + max(left, right)
        get_val(root)
        return self.res

File: leetcode/java/124BinaryTreeMaximumPathSum.java
/**
 * T: O(N) S: O(N)
 *
 * <p>We visit each node and check what's the maximum path through it. We update the global maximum
 * path and return the maximum pass through the current node using only one branch in case it is a
 * part of path with a greater sum.
 */
import java.util.ArrayList;
import java.util.Collections;

public class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {}

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

class Solution {
  private int globalMax = Integer.MIN_VALUE;

  public int maxPathSum(TreeNode root) {
    helper(root);

    return globalMax;
  }

  private int helper(TreeNode node) {
    if (node == null) {
      return Integer.MIN_VALUE;
    }
    int left = helper(node.left);
    int right = helper(node.right);
    ArrayList<Integer> pathSums = new ArrayList<>();
    pathSums.add(node.val);
    if (left > 0) {
      pathSums.add(node.val + left);
    }
    if (right > 0) {
      pathSums.add(node.val + right);
    }
    int maxOneBranchPath = Collections.max(pathSums);

    if (left > 0 && right > 0) {
      pathSums.add(node.val + left + right);
    }
    int maxTwoBranchPath = Collections.max(pathSums);
    globalMax = Math.max(globalMax, maxTwoBranchPath);

    return maxOneBranchPath;
  }
}
